src/bin.ts

```diff
@@ -403,23 +407,76 @@ function phase3(payload: BootstrapState) {
       experimentalSpecifierResolution as ExperimentalSpecifierResolution,
   });
 
-  if (preloadedConfig.options.esm) payload.shouldUseChildProcess = true;
+  // If ESM is enabled through the parsed tsconfig, stage4 should be run in a child
+  // process with the ESM loaders configured.
+  if (preloadedConfig.options.esm) {
+    payload.shouldUseChildProcess = true;
+  }
+
   return { preloadedConfig };
 }
 
+/**
+ * Determines the entry-point information from the argv and phase2 result. This
+ * method will be invoked in two places:
+ *
+ *   1. In phase 3 to be able to find a project from the potential entry-point script.
+ *   2. In phase 4 to determine the actual entry-point script.
+ *
+ * Note that we need to explicitly re-resolve the entry-point information in the final
+ * stage because the previous stage information could be modified when the bootstrap
+ * invocation transitioned into a child process for ESM.
+ *
+ * Stages before (phase 4) can and will be cached by the child process through the Brotli
+ * configuration and entry-point information is only reliable in the final phase. More
+ * details can be found in here: https://github.com/TypeStrong/ts-node/issues/1812.
+ */
+function getEntryPointInfo(
+  argvResult: NonNullable<BootstrapState['parseArgvResult']>,
+  phase2Result: NonNullable<BootstrapState['phase2Result']>
+) {
+  const { code, interactive, restArgs } = argvResult;
+  const { cwd } = phase2Result;
+
+  // Figure out which we are executing: piped stdin, --eval, REPL, and/or entrypoint
+  // This is complicated because node's behavior is complicated
+  // `node -e code -i ./script.js` ignores -e
+  const executeEval = code != null && !(interactive && restArgs.length);
+  const executeEntrypoint = !executeEval && restArgs.length > 0;
+  const executeRepl =
+    !executeEntrypoint &&
+    (interactive || (process.stdin.isTTY && !executeEval));
+  const executeStdin = !executeEval && !executeRepl && !executeEntrypoint;
+
+  /** Unresolved.  May point to a symlink, not realpath.  May be missing file extension */
+  const entryPointPath = executeEntrypoint
+    ? resolve(cwd, restArgs[0])
```

Comment:
cwd might be the wrong cwd, if the child process was forked with a different one. The cwd option has always been confusing, let me see if I can figure out why we're using it in the first place.

Reply:
I think there is an issue with the `cwd` option. It's quite ambiguous: It doesn't actually control the working directory of the invoked process (kind of surprising). It's mainly used for the script to be resolved and e.g. the tsconfig

Reply:
@cspotcode I added a new commit (which might be considered a breaking change? to me it sounds like a fix):

* The `--cwd` option now also sets the working directory properly
* The working directory is properly preserved when we transition into the child process for ESM.
* Added test scenarios for non-ESM `--cwd` and also for the scenario you outlined in this comment (fork cwd)

Reply:
I'll have to double-check.  The cwd option has never been great, IMO, and as far as I remember, is only there for backwards compatibility to support some legacy use-cases.

It makes ts-node's logic pretend to be in that cwd but without actually changing the process's cwd.  So it affects how ts-node's `--cwdMode` locates the tsconfig, and affects paths in typechecking errors.  `--cwdMode` isn't even the default anymore because it was confusing and the newer default is better.  Just goes to show, I don't love the cwd option.

As confusing as it may be, pretty sure we should not be changing it to affect `process.cwd()`.  But I'll have to double-check the old PRs where I cleaned up this logic; it was a couple years ago.  I'm not at my computer; will check later today.

Reply:
It is possible for projects to intentionally launch child processes in a different cwd, right?  Either by passing the cwd directly to `fork` options, or by changing the parent's working directory before forking.  We'll need to support that.

Reply:
Correct. Forking into different working directories should still work. I added a test for this in ESM, but should also make sure with CJS.

 We are using the cwd option in a couple of places, and naturally I also expected it to change the initial working directory, similar to how e.g. `yarn --cwd` works. This allows me to directly call TS node scripts e.g. in the `package.json` without needing more code or an extra script.

Reply:
I fixed this, and added a test. Regardless of what you decide, the additional tests can be helpful. Happy to revert the cwd changes. I believe the semantics should be that way, or the option should be removed altogether (even though it would actually be useful that way -- like `yarn --cwd`). Proposed semantics:

* `--cwd` changes the working directory before the user entry-point is invoked
* The working directory flag is omitted when forking into child processes. It's up to the user then.

Reply:
Unfortunately, I think removal makes most sense to reduce complexity.  The option predates my involvement with the project, it's confusing, and `node` nor `tsc` has an equivalent `--cwd` option.  There's also programmatic registration scenarios, where ts-node is being registered into an existing process and probably shouldn't be changing the cwd.  Could make the cwd option a CLI-only thing, but it's really just messy feature creep.

I make frequent use of `git -C` so I understand the appeal of `--cwd` in general.  But in this case, it feels like a maintenance liability to keep it around.

Can't remove till the next major, though, cuz it'd be a breaking change.

Reply:
Yes, it definitely makes some things more difficult. Not too bad IMO though.

Anyway though, I agree that the removal is a good consideration. What are the next steps for this PR?

Reply:
I think the only thing left is: entrypoint resolution in the first child uses `--cwd`, vs subsequent forked children use `process.cwd()`.

The first child process spawned by our `--esm` flag needs to obey the CLI's `--cwd` option for backwards compatibility.  (ugh annoying)  So if you are in `/project` and you run `ts-node --cwd foo bar.ts` then the entrypoint needs to be resolved to `/project/foo/bar.ts`

Subsequently forked children should behave like vanilla node and resolve their entrypoint relative to `process.cwd()` 
 If you `process.chdir('/home/user')` and do `fork('script.ts')`, the child's entrypoint needs to resolve to `/home/user/script.ts`.

---

I think this means we need to pass a boolean to child processes telling them if they are the first child spawned by `--esm`, or they are a subsequent child spawned by `fork()`.  `child-entrypoint.ts` will need to modify the `execArgv` array to pass this flag to subsequent children.  One method is a new `--is-first-child=true|false` flag, another method is to modify and re-compress the brotli payload.

---

Maybe there are other benefits to re-compressing the brotli payload: #1831 
I'm not suggesting we implement #1831 in this PR, more thinking that re-compressing the brotli payload in this PR will have benefits for future work.